### PR TITLE
Followups

### PR DESCRIPTION
**Pull request by marinaglancy**
- course backup and delete
- add help to elements
- change strings in confirmation dialogues
- remove notification "field saved"
- remove plugin sorting
- cache all categories and fields in the handler
- rename some functions
- remove "magic" template finder, instead call template from each plugin
- remove plugin.php classes and move all functions into the respective controllers
- move some behat tests from core_course into plugins and into core_customfield
- add behat generator for the custom fields, make generators for fields and categories independent from course
- move plugin-dependent code into the respective plugins
- start cleaning up providers (still not finished)

### CODE REVIEW

File: customfield/classes/handler.php [hunk 1]
```diff
@@ -584,4 +584,30 @@ public function delete_data_on_instance(int $instanceid) {
...
+     * @param field_controller $record
+     * @return string page heading
+     */
+    public function setup_edit_page(field_controller $record): string {
+        global $PAGE;
+        $url = new \moodle_url('/customfield/edit.php',
+            ['component'  => $this->get_component(), 'area' => $this->get_area(),
+                'itemid'     => $this->get_itemid(),
+                'id'         => $record->get('id'), 'type' => $record->get('type'),
+                'categoryid' => $record->get('categoryid')]);
```
Comment (tonibarbera) [inline customfield/classes/handler.php]: 12 spaces for Travis happiness

File: customfield/classes/api.php [hunk 2]
```diff
@@ -392,11 +326,13 @@ public static function move_category(category_controller $category, int $beforei
         foreach (array_values($categoriesids) as $idx => $categoryid) {
             // Use persistent class to update the sortorder for each category that needs updating.
             if ($records[$categoryid]->sortorder != $idx) {
-                $c = ($categoryid == $id) ? $category : new category_controller(0, $records[$categoryid]);
+                $c = ($categoryid == $id) ? $category : new category(0, $records[$categoryid]);
```
Comment (tonibarbera) [inline customfield/classes/api.php]: Instantiate category_controller

File: customfield/classes/api.php [hunk 3]
```diff
@@ -334,28 +336,37 @@ public static function list_categories($component, $area, $itemid): array {
...
                 $category = $categories[$data->categoryid];
             }
-            if ($data->fieldid) {
-                $fieldobj = (object)['id' => $data->fieldid, 'name' => $data->name, 'shortname' => $data->shortname,
-                    'type' => $data->type, 'configdata' => $data->configdata, 'categoryid' => $data->categoryid,
-                    'sortorder' => $data->sortorder];
+            if ($data->id) {
+                $fieldobj = new \stdClass();
+                foreach ($data as $key => $value) {
+                    if (!preg_match('/^category_/', $key)) {
```
Comment (tonibarbera) [inline customfield/classes/api.php]: :-O nice!

File: customfield/classes/handler.php (hunk 1 above)
Comment (marinaglancy) [reply customfield/classes/handler.php]: sorry, Toni, I don't understand - do you want me to change anything here?

Comment (tonibarbera) [reply customfield/classes/handler.php]: 12 spaces at 598-600, I think that should be only 12 spaces there.

**Pull request merged**